[plugins] ADD new Communication plugin

ErwanDouaille opened the pull request:
Here is a plugin communication 
The purpose of this plugin is to add ability to the users to simply send/receive data´s from sofa. Instead of creating many plugins for each implementation/hardware, we propose to regroup all of them in one communication plugin. 

Actually there is two network protocol implemented : 
- OSC (has packet size limiations ...)
- ZMQ

Improvements : 
- better thread safe implementation
- more types
- add others library implementation



______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

ErwanDouaille committed:
First commit for communication plugin

ErwanDouaille committed:
vectorData error

ErwanDouaille committed:
clean ServerCommunication, removed useless class. Still have error while parsing datawq

ErwanDouaille committed:
add refresh rate

ErwanDouaille committed:
fix data parse, working but mutex isnt the best idea ...

ErwanDouaille committed:
add server sending correct type, fix frequency problem, some code cleaning

ErwanDouaille committed:
add abstract class

ErwanDouaille committed:
update repo, error when using linked data

ErwanDouaille committed:
data copy bug almost fix

ErwanDouaille committed:
Refactor + add zmq. TODO: test zmq implementation

ErwanDouaille committed:
Remove templated class, type will now be defined from received data. Actually only the receive OSC part is working

ErwanDouaille committed:
Remove templated class, type will now be defined from received data. Actually only the receive OSC part is working

ErwanDouaille committed:
OSC is now working with automatic dataField creation and osc arguments parsing
This is not thread safe ...

ErwanDouaille committed:
changed Factory, QDisplayDataWidget, sleep to std::thread_sleep

ErwanDouaille committed:
just saving before changing everything :)

ErwanDouaille committed:
implement matrix protocol

ErwanDouaille committed:
fix error, add matrix support, add nullptr verification

damienmarchal committed:
[Communication] Add basic test class stub.

damienmarchal committed:
[Communication] Fix test that was not compiling...now it just crash on create OSC message

ErwanDouaille committed:
add examples + matrix fix

ErwanDouaille committed:
Merge branch 'sofaCommunication' of https://github.com/SofaDefrost/sofa into sofaCommunication

ErwanDouaille committed:
add documentation

ErwanDouaille committed:
Update README.md

ErwanDouaille committed:
add dev explanations

ErwanDouaille committed:
Merge branch 'master' of https://github.com/sofa-framework/sofa into sofaCommunication

ErwanDouaille committed:
Merge branch 'sofaCommunication' of https://github.com/SofaDefrost/sofa into sofaCommunication

ErwanDouaille committed:
add examples

ErwanDouaille committed:
Update README.md

ErwanDouaille committed:
Update README.md

ErwanDouaille committed:
check if data is != null for ZMQ

ErwanDouaille committed:
merge, repo + ensure data != null for ZMQ

ErwanDouaille committed:
Update zmqReceiverPublisher.cpp

ErwanDouaille committed:
fix space strings for zmq

ErwanDouaille committed:
Add crapy stuff for strings ZMQ
Merge branch 'sofaCommunication' of https://github.com/SofaDefrost/sofa into sofaCommunication

ErwanDouaille committed:
fix fetchData for createMessage OSC/ZMQ + more test

ErwanDouaille committed:
add ability to change the running variable from server communication, more tests

ErwanDouaille committed:
test separation + removed useless cout

ErwanDouaille committed:
fix blocking issue for receiver, more test. WARNING: send/receive has a segfault issue

ErwanDouaille committed:
fix ZMQ string convertion + tests

ErwanDouaille committed:
removed debug log

ErwanDouaille committed:
fix error in zmq parse stringToArgumentList

ErwanDouaille committed:
fix a bug inside stringToArgumentList, add tests

ErwanDouaille committed:
fix error in test receive of zmq. CheckThreadSafe can now be implemented

ErwanDouaille committed:
changed MyComponent to be able to test large vector and ensure it's thread safe

ErwanDouaille committed:
add/fix tests

maxime-tournier commented:
Hi, great work there!

Just to mention that since the python plugin got proper GIL handling, it is also possible to implement communication all on the python side: just start a python thread and do the message dispatch from there using the python package of your choice. From there it's fairly easy to exchange data using a `Queue` between the communication thread and e.g. a script controller (in fact that was the main reason behind implementing GIL handling).

damienmarchal commented on examples/Demos/caduceus.scn:
```diff
@@ -1,84 +1,5 @@
-            <OglModel  name="OglModel" fileMesh="mesh/SOFA_pod.obj" />
-        </Node>
-    </Node>
+<?xml version='1.0' ?>
```
Hello Erwan...replacing caduceus  is probably a mistake :)

ErwanDouaille replied:
Yeah I know. I did it for testing. (simply press ctrl+r in qtcreator then everything show up properly without modifying command line or reload scenes ... *lazy*)
Of course I have to modify this file, don´t worry ;)

ErwanDouaille commented on applications/plugins/Communication/components/serverCommunicationOSC.inl:
```diff
@@ -0,0 +1,245 @@
+    getFactoryInstance()->registerCreator("i", new DataCreator<int>());
+    getFactoryInstance()->registerCreator("s", new DataCreator<std::string>());
+
+    getFactoryInstance()->registerCreator("matrixf", new DataCreator<vector<float>>());
```
I need your help for the matrix support.
Matrix support in a nutshell, a factory is responsible to provide data depending of the type name received from the network message.
Exemple for OSC : "/aMessage f:1.0", we know the type is float (character f). If the data does not exist where we want to stock it, we ask the factory to create a data "f". In this case it will return a new DataCreator<float>().

For the matrix, a specific message is send like "/aMessage s:matrix i:1 i:3 f:1 f:2 f:3" correspond to a matrix of size 1:3 type float {1.0, 2.0, 3.0}.

For the matrix case we need a matrix wich can be resized at runtime and it have to be easily modify.

So here is the question. Which sofa type correspond to the need ?
I tried fullMatrix and didn´t succeed to read/write it. :/

ErwanDouaille committed:
first try for matrix support in OSC

ErwanDouaille committed:
first refactor of OSC

ErwanDouaille committed:
refactor OSC communication, more abstract functionnality

ErwanDouaille committed:
refactor ZMQ + matrix support

ErwanDouaille committed:
change cmake stuff

ErwanDouaille committed:
default caduceus, some test, it's not thread safe at all

damienmarchal commented:
[ci-build][with-scene-tests]

ErwanDouaille committed:
fix zmq error with matrix

damienmarchal commented:
Hello Erwan & @guparan 

Was zmq installed on the CI ? 
I installed it on centos3 and ubuntu but not on macos & windows. 
It seems there is also a dependency on liboscpack  that also needs to be installed Am I wrong ?

ErwanDouaille commented:
AFAIK, @bcarrez installed zmq on CI ubuntu (or maybe not, just before vacations). 

For liboscpack it requires a specific version : https://github.com/SofaDefrost/sofa/blob/sofaCommunication/applications/plugins/Communication/README.md
If you install liboscpack from the repository there is some bugs in it.

About windows and mac I never tested on it. Plus, I have to write a liboscpack cmake find :/

damienmarchal commented:
Thanks for the details. So It is consistent with what the CI is returning.

ErwanDouaille committed:
first try for the circular buffer and part of thread safe

ErwanDouaille committed:
I need to separate protocols from the main communication plugin

ErwanDouaille committed:
fix some issue with the value of cutoff in examples

ErwanDouaille committed:
fix a bug with matrix

ErwanDouaille committed:
some cleaning

ErwanDouaille committed:
ensure OSC examples works correctly

ErwanDouaille committed:
ensure every zmq examples works properly

ErwanDouaille committed:
clean

ErwanDouaille committed:
fix bug + no copy data, thx to @damienmarchal

ErwanDouaille committed:
clean, matrix support is back

ErwanDouaille committed:
refactor + cleaning for Circular buffer classwq

ErwanDouaille committed:
refactor the send part of osc and zmq. Now it looks similar. Needed for the implementation of an abstract function returning the datas (bufferwq)

ErwanDouaille committed:
preparing the circular buffer for the sending part

ErwanDouaille committed:
a new commit

ErwanDouaille committed:
first try for the data copy buffer sender

ErwanDouaille committed:
saving things for the windows compilation

ErwanDouaille committed:
remove pthread to std::thread ...

ErwanDouaille committed:
add oscpack cmake find

ErwanDouaille committed:
Fix for windows, thanks to @margaretkoehler and @damienmarchal

ErwanDouaille commented:
[ci-build]

ErwanDouaille committed:
comment a msg_advice throwing a segfault due to a copied data which is not well copied ...

ErwanDouaille committed:
first overview of the new tests (wip) + some fix

ErwanDouaille committed:
add windows instructions

ErwanDouaille committed:
add the sender buffer copy data

ErwanDouaille committed:
fix a bug with received message with subject only (no data)

ErwanDouaille committed:
simple cleaning

ErwanDouaille committed:
remove empty messages send. Simply added a try/catch/throw

ErwanDouaille committed:
fix stupid bug with if/else for data creation in fetchData

ErwanDouaille commented:
[ci-build]

ErwanDouaille committed:
fix part of the test + timeout for the zmq server

ErwanDouaille committed:
changed the argument

ErwanDouaille committed:
fix some tests

ErwanDouaille committed:
remove useless includes

ErwanDouaille committed:
fix osc tests

ErwanDouaille committed:
change arguments name, more user friendly. Thanks @VannesteFelix

ErwanDouaille committed:
update osc installation

ErwanDouaille committed:
update examples

ErwanDouaille committed:
remove useless sleep + fix zmq test compilation problem on windows

ErwanDouaille committed:
fix some windows stuff

ErwanDouaille committed:
fix missing libs for windows

ErwanDouaille committed:
fix a warnign

ErwanDouaille committed:
some changes for the readme

ErwanDouaille committed:
Merge branch 'sofaCommunication' of https://github.com/SofaDefrost/sofa into sofaCommunication

ErwanDouaille committed:
Merge remote-tracking branch 'origin/master' into sofaCommunication

ErwanDouaille commented:
[ci-build][with-scene-tests]

ErwanDouaille committed:
saving :s Not finished

ErwanDouaille commented:
Hey,

This PR will be pursued, by the gsoc. 
I will chat with Rupesh tonight

techytoes commented:
I think there is some issues with the CI as the communication plugin ran fine on my MacOS. What do you say @ErwanDouaille ?

techytoes committed:
adds docs about MacOS support

ErwanDouaille committed:
Merge branch 'master' into sofaCommunication

ErwanDouaille committed:
Delete blop

ErwanDouaille committed:
Merge pull request #17 from firedranzer/sofaCommunication

adds docs about MacOS support

techytoes committed:
GSOC-VRPN protocol implementation inside communication plugin

techytoes committed:
Unified scenes

techytoes committed:
improves vrpn senders and removes tracker

techytoes committed:
adds example and documentation

ErwanDouaille committed:
Merge pull request #18 from firedranzer/gsoc-rupesh

GSOC-VRPN protocol implementation inside communication plugin. From @firedranzer

techytoes committed:
GSoC - Adds Qserial protocol into Communication Plugin

techytoes committed:
Adds doc support for QSerialPort protocol

ErwanDouaille commented:
Hey @hugtalbot,

As I remember, the communication plugin is imo ready to merge. 
It still need some polish like, code cleaning, avoid some useless copy, use some boost things ... but it´s working. 

On CI, AFAIK only the core communication plugin is compiled. We don´t want to install libs for all the protocols implemented.

What you need to know is, there is still pending PR (https://github.com/SofaDefrost/sofa/pull/20 with another one) for this plugin. But they are not working and I don´t have any news from the contributor (Rupesh gsoc).
This is not a big deal because the content of the PR is an addon for the plugin Qt serial protocol.

Let me know if you have any question :)

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into sofaCommunication

# Conflicts:
#	SofaKernel/modules/SofaCore/src/sofa/core/objectmodel/Data.h
#	applications/plugins/CMakeLists.txt

guparan committed:
[Communication] CLEAN and FIX the plugin

marques-bruno committed:
Merge branch 'sofaCommunication' into gsoc-qserial-rupesh

marques-bruno committed:
Merge pull request #20 from techytoes/gsoc-qserial-rupesh

GSoC - Adds Qserial protocol into Communication Plugin

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into sofaCommunication

# Conflicts:
#	applications/plugins/CMakeLists.txt

guparan committed:
[SofaCommunication] Set default to OFF

guparan committed:
Revert "Merge pull request #20 from techytoes/gsoc-qserial-rupesh"

This reverts commit 401557559526842b446d439059113c46d8592ccd, reversing
changes made to 8b28d31813943fe5eeb7af5e11535011afb8a7a8.

guparan committed:
[Communication] CLEAN the plugin

guparan committed:
[SofaCommunication] FIX .scene-tests file

guparan committed:
[Communication] FIX Oscpack finding

guparan committed:
[Communication] FIX ZMQ and Oscpack finding

guparan committed:
[Communication] FIX OSCPACK_FOUND in CMakeLists

guparan committed:
[Communication] FIX Oscpack finding (2)

guparan committed:
[Communication] FIX build on Windows

guparan committed:
[Communication] CLEAN ZeroMQ finding

The CMake package is called ZeroMQ, not ZMQ.

guparan committed:
[SofaMacros] FIX sofa_install_libraries with imported targets

guparan committed:
[Communication] ADD dynamic .scene-tests file

guparan committed:
[Communication] FIX case issues

find_package_handle_standard_args sets only <PACKAGENAME>_FOUND and not <PackageName>_FOUND

guparan commented:
[ci-build][with-all-tests][force-full-build]

hugtalbot committed:
Fix warning reorder

guparan commented:
[ci-build][with-all-tests]

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into sofaCommunication

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into sofaCommunication

# Conflicts:
#	SofaKernel/SofaFramework/SofaMacros.cmake

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into sofaCommunication

# Conflicts:
#	SofaKernel/modules/Sofa.Config/cmake/SofaMacrosInstall.cmake

guparan commented:
[ci-build][with-all-tests]

guparan committed:
[Communication] FIX ExecParams with opaque API

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into sofaCommunication

guparan committed:
[Communication] Remove FullMatrix<int> instanciations

guparan commented:
Updated one last time and moved to https://github.com/sofa-framework/Communication
Feel free to open issues and pull-requests there.
